### DISCUSSION TITLE
Spatial condition on DirichletBC function

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#20275] [Q&A Getting Started] [answered]
Hi everyone,

I recently started using MOOSE to do some elastic deformation. I would like to prescribe a short (in time) and localized (in space) change in the Dirichlet boundary condition on one of the sides of my domain. I used inline functions before to add temporal and spatial constraints to constant boundary conditions. A generic example:
```
[BCs]
  [BC_1]
     type = FunctionDirichletBC
      variable = my_variable
      boundary = 'my_boundary'
      function = 'if( (x>1.) & (x<2.) & (t>1.), 1., 0.)'
  []
[]
```

But now I have a forcing with values through time stored in a file that I load with a function:
```
[Functions]
  [forcing]
    type = PiecewiseLinear
    data_file = my_forcing.csv
    scale_factor = 1.0
    format = 'columns'
  []
[]
```

And would like to apply it when specific spatial conditions are met. I initially thought the following could maybe be possible in `BC_1`:
``function = 'if( (x>1.) & (x<2.), forcing, 0)`` or ``function = 'if( (x>1.) & (x<2.), forcing(t), 0)`` but this is not the case. 

All the input files are very useful and helped me a lot to setup my app but in this case I am struggling a bit to find the right strategy.

Maybe the/a proper way to do it would be to, at the mesh generation step, give a different ` blockID` to the elements where I want to prescribe this forcing and then use block ids in the input file?

Any help would be much appreciated! Thank you a lot in advance!

P.S. I've been enjoying MOOSE a lot so far, being a Libmesh user this is making everything much easier and more efficient! Thank you a lot for all your work! :+1:

**Comment by wehrad** [comment]
So I think I found a solution: generate a new side set with a given id that I can then use in the different objects. Just like here:

(see @GiudGiud's comment below with the different possible options!) 

https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/test/tests/mesh_modifiers/add_side_sets_from_bounding_box/multiple_boundary_ids_3d.i#L19-L26

However I'm still interesting to see if you think it's a good/the best solution?

**Reply by GiudGiud** [reply]
For the space restriction, this is a good idea. Then do the time restriction with a function.

The best solution would be a ParsedDirichletBC (similar to ParsedAux or ParsedPostprocessor) in my opinion. We havent developed that one yet.

**Reply by wehrad** [reply]
Thank you a lot for your answer!

**Reply by wehrad** [reply]
For the time restriction, I'm not sure how to proceed since I already have a `forcing` function (hence a variable forcing with time). Like in my post but with x being time:
> And would like to apply it when specific spatial conditions are met. I initially thought the following could maybe be possible in BC_1:
> function = 'if( (x>1.) & (x<2.), forcing, 0) or function = 'if( (x>1.) & (x<2.), forcing(t), 0) but this is not the case.

**Reply by GiudGiud** [reply]
x is time?
So the condition (in fake code) is:
if( (t>1.) & (t<2.), forcing(t), 0)  ?

**Reply by wehrad** [reply]
Yes! That was my initial idea.

**Reply by GiudGiud** [reply]
A few options:

1) add a scaling factor argument (currently does not exist) to the FunctionDirichletBC https://mooseframework.inl.gov/source/bcs/FunctionDirichletBC.html
(This has a scale_factor in the code if you want an example: https://mooseframework.inl.gov/source/functions/CompositeFunction.html)
then use the Control system to modify it with the step function : https://mooseframework.inl.gov/source/controls/RealFunctionControl.html

2) use the RealFunctionContrl https://mooseframework.inl.gov/source/controls/RealFunctionControl.html directly with the Function to control a DirichletBC
https://mooseframework.inl.gov/source/bcs/DirichletBC.html

3) Make your own DirichletBC that loads the CSV and uses it, so merge PieceWiseLinear and DirichletBC for your app. 

4) Use a https://mooseframework.inl.gov/source/functions/CompositeFunction.html with a step function multiplier, for the FunctionDirichletBC

2 and 4 are no-code

**Reply by wehrad** [reply]
Thank you a lot for all those options! At the end I decided to have the temporal conditions directly prescribed in my forcing csv file as they might get quite complicated at some point, and would result in very long temporal conditions. But I will for sure need the options you just listed in a short moment!

In parallel, more than controlling the value associated with a given DirichletBC, I would like to completely enable/disable a Dirichlet BC at given times (in my case, releasing the boundary over a short period of time). And for that I think I am really stuck now. Again, any tips would be much appreciated!

Thank you a lot for your help!

**Reply by wehrad** [reply]
> In parallel, more than controlling the value associated with a given DirichletBC, I would like to completely enable/disable a Dirichlet BC  > at given times (in my case, releasing the boundary over a short period of time).

I think the controls system is the/a way to go (as used e.g. here https://github.com/idaholab/moose/discussions/20208). Do you agree?
Thank you a lot for your help!

**Comment by GiudGiud** [comment, marked as answer]
Control system is the way to go here. Should work fine on a DirichletBC

**Reply by wehrad** [reply]
Thank you a lot!
